gk/main.c

```diff
@@ -548,6 +555,67 @@ gk_setup_rss(struct gk_config *gk_conf)
 	return ret;
 }
 
+static void
+gk_arp_and_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	struct ether_addr *nexthop_mac = arg;
+
+	/* TODO Add concurrency control by sequential lock on nexthop_mac. */
+	if (!map->stale)
+		ether_addr_copy(&map->ha, nexthop_mac);
+	else
+		/* 00:00:00:00:00:00 represents invalid MAC address. */
+		memset(nexthop_mac, 0, sizeof(*nexthop_mac));
+
+	*pcall_again = true;
+}
+
+static int
+gk_hold_arp_and_nd(struct gk_lpm *ltbl, unsigned int lcore_id)
+{
+	unsigned int i;
+	int ret;
+	struct gk_policy *policies = ltbl->policy_tbl;
+	for (i = 0; i < ltbl->num_policies; i++) {
+		struct ipip_tunnel_info *tunnel;
+
+		if (policies[i].action != GK_FWD_GT)
+			continue;
+
+		tunnel = &policies[i].grantor_tunnel;
+		if (tunnel->flow.proto == ETHER_TYPE_IPv4) {
+			struct in_addr *ipv4_dst;
+			if (policies[i].is_gt_in_back_net)
+				ipv4_dst = (struct in_addr *)
+					&tunnel->flow.f.v4.dst;
+			else
+				ipv4_dst = &policies[i].gw->ip.v4;
+
+			ret = hold_arp(gk_arp_and_nd_req_cb, (void *)
```

Comment:
(1) The encapsulation function should not add the Ethernet header. This way we can compose the Ethernet header copying it from a cache, or by setting his fields. The final way to add the Ethernet header will vary, but it would not come from `struct ipip_tunnel_info`.

(2) Consider that the `struct gk_policy` of the action `GK_FWD_GRANTOR` has inside its union `struct ipip_tunnel_info`. The second lookup of the LPM table informs that the Grantor server is not local, so it returns a `struct gk_policy` of the action GK_FWD_GATEWAY. Notice that the LLS will resolve the `struct gk_policy` of the action GK_FWD_GATEWAY, and never update the MAC address in `struct ipip_tunnel_info` of the first `struct gk_policy`. Did you follow?

(3) That's the right place.